### PR TITLE
Inverse operability analysis of ethanol–water distillation with surrogate validation

### PR DESCRIPTION
**Pull request by Spogis**
This PR finalizes a surrogate-based inverse operability analysis for an ethanol–water distillation column.

The notebook was refined to ensure full consistency between AIS, AOS, DOS, and DIS*, with the inverse problem formulated using product-level specifications. The MLP surrogate was validated against the DWSIM first-principles model, showing excellent agreement and a large computational speedup.

Overall, this update improves conceptual clarity, result presentation, and robustness of the operability case study, making it suitable for research and teaching purposes.

### CODE REVIEW

**Comment by rafaelrheil** [comment]
I tested the mlp_ethanol_water_distillation.ipynb notebook and found a small issue. In the dc_problem function, manager.CalculateFlowsheet4(myflowsheet) is called twice on consecutive lines:

errors = manager.CalculateFlowsheet4(myflowsheet)
errors = manager.CalculateFlowsheet4(myflowsheet)  # ← Remove this duplicate

The duplicate call produces NaN outputs when running the "Quick Sanity Check" cell. Removing one of these lines fixes the issue.

Also, in the inverse mapping section, using warmstart=True produced misaligned results between the First-Principles (DWSIM) and Surrogate (MLP) models in the overlay plots (DIS* and DOS*) for me. Many blue circles don't coincide with red X markers.

Setting warmstart=False resolved this and I got the same plot as in the original output.

The rest of the notebook and the other one looks good to me.

**Comment by Spogis** [comment]
Thanks Rafael!

I'll make these modifications and send a new PR.

Best regards,

Nicolas

**Comment by rafaelrheil** [comment]
Tested 4f454a7, everything now runs correctly on my end and overlay plots show good agreement between surrogate and first-principles in both Jupyter notebooks. Thanks Nicolas!

**Comment by victoraalves** [comment]
Hi,

Although this example is conceptually correct, in my opinion, it does not bring anything new to the Examples Gallery. Using surrogates for process operability calculations has been a reality since my [CACE publication (4 years ago)](https://doi.org/10.1016/j.compchemeng.2022.107835), and we already have more than one example from yourself, and from @rafaelrheil, [namely the use of surrogates for operability analysis in the shower problem example](https://codes-group.github.io/opyrability/examples_gallery/gp_surrogate_shower.html),[ the membrane reactor using a surrogate model](https://codes-group.github.io/opyrability/examples_gallery/gp_surrogate_dma-mr.html) and [your excellent heat exchanger example](https://codes-group.github.io/opyrability/examples_gallery/mlp_DWSIM_heat_exchanger.html).

My other concern is advancing the state-of-the-art of opyrability. For example , https://github.com/CODES-group/opyrability/pull/33, which I am currently reviewing, is a great example of adding OMLT/IDAES support at a core, fundamental level of the `opyrability` codebase, and not just a simple example that recycles functionality of the tool in a repeated way.


I believe we should take a step back and rethink what are the priorities of the tool, @savannahsakhai , @fvlima-codes. I want to **not only** focus on having a diverse set of examples in our example gallery (which is extremely relevant, and I appreciate them a lot!), but also develop the tool in a way that it's relevant in the long run.

I'd suggest holding this PR until we decide what are the next steps forward.

**Comment by Spogis** [comment]
> Hi,
> 
> Although this example is conceptually correct, in my opinion, it does bring anything new to the Examples Gallery. Using surrogates for process operability calculations has been a reality since my [CACE publication (4 years ago)](https://doi.org/10.1016/j.compchemeng.2022.107835), and we already have more than one example from yourself, and from @rafaelrheil, [namely the use of surrogates for operability analysis in the shower problem example](https://codes-group.github.io/opyrability/examples_gallery/gp_surrogate_shower.html),[ the membrane reactor using a surrogate model](https://codes-group.github.io/opyrability/examples_gallery/gp_surrogate_dma-mr.html) and [your excellent heat exchanger example](https://codes-group.github.io/opyrability/examples_gallery/mlp_DWSIM_heat_exchanger.html).
> 
> My other concern is advancing the state-of-the-art of opyrability. For example , #33, which I am currently reviewing, is a great example of adding OMLT/IDAES support at a core, fundamental level of the `opyrability` codebase, and not just a simple example that recycles functionality of the tool in a repeated way.
> 
> I believe we should take a step back and rethink what are the priorities of the tool, @savannahsakhai , @fvlima-codes. I want to **not only** focus on having a diverse set of examples in our example gallery (which is extremely relevant, and I appreciate them a lot!), but also develop the tool in a way that it's relevant in the long run.
> 
> I'd suggest holding this PR until we decide what are the next steps forward.

Thanks again for the thoughtful and strategic feedback — I genuinely appreciate the broader perspective you’re bringing to this discussion.

That said, I still believe this specific case is a relevant contribution to the Examples Gallery, and I’d like to clarify why I see it as distinct from the existing surrogate-based operability examples.

1)Higher complexity / stronger numerical instability (distillation column)
Compared to the current examples, this column case is significantly more challenging from a convergence and robustness standpoint. We are dealing with a distillation column exhibiting strong nonlinearities and solution instabilities. In practice, this resulted in noisy and unreliable derivative information — precisely the type of scenario where methods that work well on smoother problems begin to struggle.

2) A derivative-free optimization path (Nelder–Mead) driven by necessity
Due to the noise in the derivatives, IPOPT was not a viable option in this case. This was not a modeling preference, but a numerical constraint imposed by the problem structure. As a result, we had to adopt a derivative-free strategy (Nelder–Mead). I believe this is valuable for the gallery because it documents a realistic failure mode and a practical workaround that users are likely to encounter in complex flowsheet-based problems.

3) More complex surrogate structure (higher-capacity MLPs)
The surrogate models used here are not lightweight regressors. They are more complex MLP architectures, which introduce additional challenges in terms of smoothness, conditioning, and sensitivity behavior. This affects optimization performance and stability, and therefore meaningfully changes the operability workflow compared to simpler surrogate cases.

4) A stronger focus on process control and operability under dynamic sensitivity
This example is also more explicitly connected to process control considerations. The distillation column case is not only about steady-state feasibility or geometric operability regions, but about how input variations propagate through a highly sensitive, tightly coupled system. In columns, small perturbations in reflux ratio, feed conditions, or tray efficiencies can lead to amplified composition responses. From a control perspective, this makes the operability region more fragile and highlights the importance of robustness against disturbances.

By combining surrogate modeling with derivative-free optimization in this context, the example bridges operability analysis with practical control-oriented concerns — something that is less emphasized in the existing examples.

That said, I fully agree with the broader point about advancing the core of the tool and ensuring long-term relevance. Contributions like #33, especially with deeper OMLT/IDAES integration, are strategically important for the evolution of opyrability at a fundamental level.

So I’m completely comfortable leaving the decision with Victor. If the maintainers prefer to pause on new examples until priorities are clarified, I support that direction. I just wanted to clearly explain why, from my perspective, this PR introduces a different stress-test dimension — combining distillation instability, derivative-free optimization, higher-capacity MLP surrogates, and a stronger control-oriented framing — which I believe is not yet represented in the current gallery.

**Comment by victoraalves** [comment]
@Spogis I have a slightly different opinion. Please let me elaborate on why this is the case by responding to your last comment, if I may:

> 1)Higher complexity / stronger numerical instability (distillation column) 
> Compared to the current examples, this column case is significantly more challenging from a convergence and robustness standpoint. We are dealing with a distillation column exhibiting strong nonlinearities and solution instabilities. In practice, this resulted in noisy and unreliable derivative information — precisely the type of scenario where methods that work well on smoother problems begin to struggle.

The polytopic AOS regions shown on your notebook are literally convex. They do not show any sign of strong non-convexity or nonlinearities. Plus, optimization and inverse problems of distillation systems can and actually are performed using gradient-based solver for decades. While I agree it can be challenging, the variables you selected for your AOS does to show this:

image


> 2. A derivative-free optimization path (Nelder–Mead) driven by necessity
> Due to the noise in the derivatives, IPOPT was not a viable option in this case. This was not a modeling preference, but a numerical constraint imposed by the problem structure. As a result, we had to adopt a derivative-free strategy (Nelder–Mead). I believe this is valuable for the gallery because it documents a realistic failure mode and a practical workaround that users are likely to encounter in complex flowsheet-based problems.

I believe you mean numerical noise, which can be, among other things, solved by for example (i) tightening solver tolerances in the process simulator, (ii) using an integrated EO approach, and (iii) using a surrogate that is differentiable and can give derivative information itself, or (iv) use opyrability's capability of automatic differentiation and train a machine learning model in for example, JAX. Opyrability already does (iv) natively. All of these solutions are more modern than what is claimed in your response. Plus, if you check the code:

```
# Robust MLP for inverse operability (smooth + regularized + early stopping)
mlp = MLPRegressor(
    hidden_layer_sizes=(32, 32),
    activation="tanh",
    solver="adam",
    alpha=1e-3,
    learning_rate_init=5e-4,
    max_iter=8000,
    early_stopping=True,
    validation_fraction=0.15,
    n_iter_no_change=30,
    tol=1e-6,
    random_state=42
)
```

It's diffcult to understand the need of such a large NN architecture (2 -> 32 -> 32 -> 2) for such a low dimensional example, that yields a convex polytope as shown by your results in your jupyter notebook.


> 3. More complex surrogate structure (higher-capacity MLPs)
> The surrogate models used here are not lightweight regressors. They are more complex MLP architectures, which introduce additional challenges in terms of smoothness, conditioning, and sensitivity behavior. This affects optimization performance and stability, and therefore meaningfully changes the operability workflow compared to simpler surrogate cases.

Fair point, but hard to believe that this example is too hard in a simple 2x2 case study. Again, unless your AOS envelope is unintentionally hiding any folds/bifurcations, (that can happen, not your fault at all) it does not seem to be the case (e.g., Delaunay triangulation of the points seems to be quite well-behaved).


>4.  A stronger focus on process control and operability under dynamic sensitivity
> This example is also more explicitly connected to process control considerations. The distillation column case is not only about steady-state feasibility or geometric operability regions, but about how input variations propagate through a highly sensitive, tightly coupled system. In columns, small perturbations in reflux ratio, feed conditions, or tray efficiencies can lead to amplified composition responses. From a control perspective, this makes the operability region more fragile and highlights the importance of robustness against disturbances.

Your AIS variables are the (i) bottoms molar flow rate and the (ii) distillate composition, while your AOS variables are the (i) distillate molar flow rate and the (ii) reboiller duty. It is indeed a great idea to bring the simultaneous design and control objectives of process operability, this unfortunately is not quite the case here:

You have control variables (i.e., composition) defined in the input space. Although this is totally fine as you can consider that you have the ability to freely change its set point if there is a controller that can achieve that set point in the first place, it would be more interesting, in my opinion, to have it in the AOS. Then one would see the nonlinear behavior of the separation system, a result of VLE, manifested in the composition now treated as a free variable.

Your second AIS variable is the bottoms flow rate, and one of your AOS variables is the distillate flow rate. These can be solved by a simple molar balance, or in other words, one output of your MLP is a linear function of one of your features, in ML lingo. 


This separation example is a great one. However, in my opinion, it  needs to be thoroughly reviewed in terms of variables in the AIS/AOS spaces if your original intent was to bring the controls portion of operability analysis, as well as to make it sufficiently different from the other ones. As is, it looks like the other examples to me and it does not capture the features of what you are claiming.

@savannahsakhai, @fvlima-codes, thoughts?
